<a href="https://colab.research.google.com/github/kaholict/Karpov-Course/blob/main/K_C_lesson_2_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/bookings.csv', sep=';')

In [ ]:
df_head = df.head(7)
df_head

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [ ]:
df.columns = [name.lower().replace(' ', '_') for name in df.columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 21 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   hotel                      119390 non-null  object 
 1   is_canceled                119390 non-null  int64  
 2   lead_time                  119390 non-null  int64  
 3   arrival_full_date          119390 non-null  object 
 4   arrival_date_year          119390 non-null  int64  
 5   arrival_date_month         119390 non-null  object 
 6   arrival_date_week_number   119390 non-null  int64  
 7   arrival_date_day_of_month  119390 non-null  int64  
 8   stays_in_weekend_nights    119390 non-null  int64  
 9   stays_in_week_nights       119390 non-null  int64  
 10  stays_total_nights         119390 non-null  int64  
 11  adults                     119390 non-null  int64  
 12  children                   119386 non-null  float64
 13  babies                     11

* Hotel – тип отеля (City Hotel или Resort Hotel)  
* Is canceled – бронирование было отменено (1) или нет (0); не отмененное *  считается успешным
* Lead time – количество дней, прошедших между датой бронирования и датой прибытия  
* Arrival full date – полная дата прибытия
* Arrival date year – год прибытия  
* Arrival date month – месяц прибытия  
* Arrival date week number – номер недели прибытия
* Arrival date day of month – день прибытия
* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле
* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле
* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)
* Adults – число взрослых
* Children – число детей
* Babies – число младенцев 
* Meal – выбранный тип питания
* Country – страна происхождения клиента
* Reserved room type – тип зарезервированного номера
* Assigned room type – тип полученного номера (может отличаться от забронированного)
* Customer type – тип бронирования
* Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; * No-Show - клиент не зарегистрировался и сообщил администрации отеля причину
* Reservation status date – дата обновления статуса

**Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.**

In [ ]:
df.head()

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

In [ ]:
df.query("is_canceled == 0").groupby('country', as_index=False).agg({'is_canceled': 'count'}).sort_values('is_canceled', ascending=False).head()

country  is_canceled
125     PRT        21071
57      GBR         9676
54      FRA         8481
50      ESP         6391
42      DEU         6069

Наибольшее число успешных бронированией было сделано из стран PRT, GBR, FRA, ESP, DEU

***На сколько ночей в среднем бронируют отели разных типов?***

In [ ]:
df.query("is_canceled == 0").groupby('hotel', as_index=False).agg({'stays_total_nights': 'mean'}).round(2)

hotel  stays_total_nights
0    City Hotel                2.92
1  Resort Hotel                4.14

***Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?***

In [ ]:
len(df.query("reserved_room_type != assigned_room_type"))

14917

Встретилось 14917 несовпадений типов номеров при брони и при заселении.

***Проанализируйте даты запланированного прибытия.   
– На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?  
– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов***

In [ ]:
month_2016 = df.query("arrival_date_year == 2016 and is_canceled == 0").groupby('arrival_date_month', as_index=False).agg({'is_canceled': 'count'})
month_2016.arrival_date_month[month_2016.is_canceled.idxmax()]

'October'

In [ ]:
month_2017 = df.query("arrival_date_year == 2017 and is_canceled == 0").groupby('arrival_date_month', as_index=False).agg({'is_canceled': 'count'})
month_2017.arrival_date_month[month_2017.is_canceled.idxmax()]

'May'

В 2016 году самым популярным месяцем был октябрь, однако в 2017 им оказался май.

In [ ]:
(
    df
      .query('hotel == "City Hotel" and is_canceled == 1')
      .groupby('arrival_date_year')
      .arrival_date_month
      .value_counts()
)

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

*    В 2015 больше всего отмен было в Сентябре
*    В 2016 больше всего отмен было в Октябре
*    В 2017 больше всего отмен было в Мае


**Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?**

In [ ]:
df[['adults', 'children', 'babies']].describe()

adults       children         babies
count  119390.000000  119386.000000  119390.000000
mean        1.856403       0.103890       0.007949
std         0.579261       0.398561       0.097436
min         0.000000       0.000000       0.000000
25%         2.000000       0.000000       0.000000
50%         2.000000       0.000000       0.000000
75%         2.000000       0.000000       0.000000
max        55.000000      10.000000      10.000000

Наибольшее среднее значние имеет переменная 'adults'

***Создайте колонку total_kids, объединив children и babies.  
Отели какого типа в среднем пользуются большей популярностью у клиентов с детьми?***

In [ ]:
df[['children', 'babies']].fillna(0)

children  babies
0            0.0       0
1            0.0       0
2            0.0       0
3            0.0       0
4            0.0       0
...          ...     ...
119385       0.0       0
119386       0.0       0
119387       0.0       0
119388       0.0       0
119389       0.0       0

[119390 rows x 2 columns]

In [ ]:
df['total_kids'] = df.children + df.babies

In [ ]:
df.groupby('hotel').agg({'total_kids': 'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

Отели типа City Hotel в среднем пользуются большей популярностью у клиентов с детьми.

***Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False.  
Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate).  
Укажите, среди какой группы показатель выше.***

In [ ]:
df['has_kids'] = [True if total_kid > 0 else False for total_kid in df.total_kids]
# ИЛИ ТАК
# df['has_kids'] = df.total_kids > 0
df['has_kids']

0         False
1         False
2         False
3         False
4         False
          ...  
119385    False
119386    False
119387    False
119388    False
119389    False
Name: has_kids, Length: 119390, dtype: bool

In [ ]:
no_kids_churn = df.query('has_kids == False and is_canceled == 1').shape[0] / df.query('has_kids == False').shape[0]
no_kids_churn = round(no_kids_churn * 100, 2)
no_kids_churn

37.22

In [ ]:
has_kids_churn = df.query('has_kids == True and is_canceled == 1').shape[0] / df.query('has_kids == True').shape[0]
has_kids_churn = round(has_kids_churn * 100, 2)
has_kids_churn

34.92

In [ ]:
import pandas as pd

url = 'https://drive.google.com/file/d/0B6GhBwm5vaB2ekdlZW5WZnppb28/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
df = pd.read_csv(path)

HTTPError: ignored